boa_engine/src/string.rs

```diff
@@ -358,57 +766,71 @@ impl JsString {
 
-        let this = Self {
-            inner: Inner::concat_array(&[x, y]),
-            _marker: PhantomData,
-        };
+        let inner = Inner::concat_array(&[x, y]);
 
-        if this.len() <= MAX_CONSTANT_STRING_LENGTH {
-            if let Some(constant) = CONSTANTS.with(|c| c.get(&this).cloned()) {
+        if unsafe { inner.as_ref() }.len <= MAX_CONSTANT_STRING_LENGTH {
+            if let Some(constant) =
+                CONSTANTS.with(|c| c.get(unsafe { inner.as_ref() }.as_str()).cloned())
+            {
                 return constant;
             }
         }
 
-        this
+        Self {
+            // Safety: We already know it's a valid heap pointer.
+            inner: unsafe { Flag::new_heap(inner) },
+            _marker: PhantomData,
+        }
     }
 
     /// Concatenate array of string.
     pub fn concat_array(strings: &[&str]) -> Self {
-        let this = Self {
-            inner: Inner::concat_array(strings),
-            _marker: PhantomData,
-        };
+        let inner = Inner::concat_array(strings);
 
-        if this.len() <= MAX_CONSTANT_STRING_LENGTH {
-            if let Some(constant) = CONSTANTS.with(|c| c.get(&this).cloned()) {
+        if unsafe { inner.as_ref() }.len <= MAX_CONSTANT_STRING_LENGTH {
+            if let Some(constant) =
+                CONSTANTS.with(|c| c.get(unsafe { inner.as_ref() }.as_str()).cloned())
+            {
                 return constant;
             }
         }
 
-        this
+        Self {
+            // Safety: We already know it's a valid heap pointer.
+            inner: unsafe { Flag::new_heap(inner) },
+            _marker: PhantomData,
+        }
     }
 
     /// Return the inner representation.
     #[inline]
-    fn inner(&self) -> &Inner {
-        unsafe { self.inner.as_ref() }
+    fn inner(&self) -> InnerKind<'_> {
+        // Check the first bit to 1.
+        if self.inner.is_static() {
+            // Safety: We already checked.
+            InnerKind::Static(unsafe { self.inner.get_static_unchecked() })
+        } else {
+            // Safety: We already checked.
+            InnerKind::Heap(unsafe { self.inner.get_heap_unchecked().as_ref() })
+        }
     }
 
     /// Return the JavaScript string as a rust `&str`.
     #[inline]
     pub fn as_str(&self) -> &str {
-        let inner = self.inner();
-
-        unsafe {
-            let slice = std::slice::from_raw_parts(inner.data.as_ptr(), inner.len);
-            std::str::from_utf8_unchecked(slice)
+        match self.inner() {
+            InnerKind::Heap(inner) => inner.as_str(),
+            InnerKind::Static(inner) => inner,
         }
     }
 
     /// Gets the number of `JsString`s which point to this allocation.
     #[inline]
     pub fn refcount(this: &Self) -> usize {
-        this.inner().refcount.get()
+        match this.inner() {
+            InnerKind::Heap(inner) => inner.refcount.get(),
+            InnerKind::Static(_inner) => 0,
+        }
```

Comment:
This should probably return `Option<usize>` instead, since it doesn't really makes sense to return 0 on static strings.

Reply:
The most appropriate type is `Option<NonZeroUsize>`, but I think we don't need it.